### PR TITLE
client: save receipts and implement getLogs, getTxReceipt, getTxByHash

### PR DESCRIPTION
**Pull request by ryanio**
This PR adds receipt and log saving functionality to the client.

A new leveldb instance is added called `metaDB` which stores this data and can be used for more data/caching/indexes in the future.

The new `ReceiptsManager` handles receipt and log saving and lookup, along with the `txHash -> [blockHash, txIndex]` index that is limited by `--txLookupLimit` (`default = 2350000 blocks = about one year, 0 = entire chain`)

Enable with `--saveReceipts` and use with `--rpc` to enable `eth_getLogs`, `eth_getTransactionReceipt` and `eth_getTransactionByHash` endpoints.

### CODE REVIEW

**Comment by ryanio** [comment]
Here's a sample output from early mainnet sync, picked a tx from block 671204 and queried it by hash and returned the tx:

Screen_Shot_2021-11-07_at_4 06 23_PM

I tried to add extensive tests and coverage and will check the coverage report to see if I missed anything:

```sh
TAP version 13
# eth_getLogs: call with valid arguments
ok 1 should return the correct logs
ok 2 should return the correct logs (filter by address)
ok 3 should return the correct logs (filter by topic - empty means anything)
ok 4 should return the correct logs (filter by topic - exact match)
ok 5 should return the correct logs (filter by topic - exact match for second topic)
ok 6 should return the correct logs (filter by topic - A or B in first position)
ok 7 should return the correct logs (filter by topic - null means anything)
ok 8 should return the correct logs (filter by blockHash)
# eth_getLogs: call with invalid params
ok 9 should return error object with error code and message
ok 10 should return error object with error code and message
ok 11 should return error object with error code and message
ok 12 should return error object with error code and message
ok 13 should return error object with error code and message
ok 14 should return empty logs
ok 15 should return empty logs

# eth_getTransactionReceipt: call with legacy tx
ok 1 should return the correct tx
# eth_getTransactionReceipt: call with 1559 tx
ok 2 should return the correct tx
# eth_getTransactionReceipt: call with unknown tx hash
ok 3 should return null

# eth_getTransactionByHash: call with legacy tx
ok 1 should return the correct tx
ok 2 should return null when past txLookupLimit
# eth_getTransactionByHash: call with 1559 tx
ok 3 should return the correct tx type
ok 4 should return the correct tx when txLookupLimit=0
# eth_getTransactionByHash: call with unknown tx hash
ok 5 should return null
```

**Comment by holgerd77** [comment]
Phew.

What an amount of work. 🙂

**Comment by holgerd77** [comment]
I'll give this a short do-no-merge/blocked, so that we are not accidentally merging this in into the current round of releases.

**Comment by acolytec3** [comment]
All looks good based on my testing.  I can go ahead and approve or we can hold off until after the releases and then rebase and test again.

**Comment by ryanio** [comment]
> All looks good based on my testing. I can go ahead and approve or we can hold off until after the releases and then rebase and test again.

thanks! no need to approve yet, I'll add a commit to update the branch to master once the new releases branch is merged, then an approval would be great :)

**Comment by holgerd77** [comment]
Let's please do a more in-depth review here. It can't be that there is no comment to make on such a heavy code (change) base. So this needs at least a 1-2 hour review to be solid.

File: packages/client/test/rpc/helpers.ts
Comment (acolytec3) [inline packages/client/test/rpc/helpers.ts:249]:
```suggestion
 * Randomly generated acount for testing purposes (signing txs, etc.)
```

**Review by acolytec3 — CHANGES_REQUESTED** [review]
I've tested the code pretty extensively and verified responses against Infura so looks good minus the one nit on spelling.

File: packages/vm/src/runBlock.ts
Comment (holgerd77) [inline packages/vm/src/runBlock.ts:440]: What is this change here?

**Comment by holgerd77** [comment]
@acolytec3 ok, just want to make sure that this is not waved through too light-heartedly. 🙂 

Generally this can now be merged once ready with the releases now out, have removed the respective "blocked" and "do-not-merge" labels.

**Comment by acolytec3** [comment]
>@acolytec3 ok, just want to make sure that this is not waved through too light-heartedly. slightly_smiling_face

>Generally this can now be merged once ready with the releases now out, have removed the respective "blocked" and "do-not-merge" labels.

Just to clarify, I skimmed through the code and did most of my validation by running the client and then validating responses against the same request made to Infura's RPC endpoint.  I can do a more thorough review of the code line by line in the next day or two as life allows.

File: packages/vm/src/runBlock.ts
Comment (ryanio) [reply packages/vm/src/runBlock.ts:440]: ah I wrote a comment about this line to narrate why but I must not have submitted it.

I was pretty surprised to find `rlp.encode(Object.values(receipt))` was working to rlp the receipt - `Object.values` converts the object's properties to an array and I guess we were lucky enough that the array order was exactly matching to the yellow paper specification (`[stateRoot, gasUsed, logsBloom, logs]`, post-Byzantium `stateRoot` replaced with `status`). I think being more explicit here is better for readability/understandability/upgradability so I made the update.

File: packages/client/test/rpc/helpers.ts
Comment (ryanio) [reply packages/client/test/rpc/helpers.ts]: thx!

**Comment by holgerd77** [comment]
Updated this branch.

**Comment by ryanio** [comment]
notes from our call via questions from @gabrocheleau:

`GET_LOGS_LIMIT` is already implemented (constant: 10,000)

Likely we also want to add these two to reduce DDoS vectors:

`GET_LOGS_BLOCK_RANGE_LIMIT` (suggest: 2,000)
`GET_LOGS_MAX_SIZE_MB` (suggest: 20MB)  (for devp2p layer)

I'll add a commit to add them today.

**Comment by ryanio** [comment]
this is ready for final review/approval and merge, I can follow up with more sophisticated reorg functionality in a future PR. thanks!

**Review by acolytec3 — APPROVED** [review]
LGTM!  I added a test for `PooledTransactions` (admittedly not a core part of the PR but where code was touched) to help bump the code coverage but looks like it's still not at 82%.  I'd be happy to see it merged though and we can work on coverage with future PRs.

**Comment by ryanio** [comment]
ah great thanks! that's a nice idea, i just wrote some more tests for missed coverage as well, adding a commit now

**Review by gabrocheleau — APPROVED** [review]
Looks good to me

File: packages/client/bin/cli.ts
Comment (ryanio) [reply packages/client/bin/cli.ts:427]: thanks for the review!
they're newline separators for the `--helprc` response

**Comment by ryanio** [comment]
ok, will merge this in now with 2 approvals, thanks everyone

**Comment by holgerd77** [comment]
Nice! 🙂

**Pull request merged**
